### PR TITLE
domain: stop schema validator to prohibit DML executing when tidb lost session of etcd

### PR DESCRIPTION
**Pull request by winkyao**
### What problem does this PR solve? 
The etcd is responsible for schema synchronization, we should ensure there is at most two diffrent schema version in the TiDB cluster, to make the data/schema be consistent. 

If we lost connection/session to etcd, the cluster will treats this TiDB as a down instance, and etcd will remove the key of `/tidb/ddl/all_schema_versions/tidb-id`. 

Say the schema version now is 1, the owner is changing the schema version to 2, it will not wait for this down TiDB syncing the schema, then continue to change the TiDB schema to version 3. Unfortunately, this down TiDB schema version will still be version 1. And version 1 is not consistent to version 3. 

So we need to stop the schema validator to prohibit the DML executing when we lost session of etcd.

This PR revert partial changes of https://github.com/pingcap/tidb/pull/7810

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8441)

### CODE REVIEW

**Comment by winkyao** [comment]
@zimulala @tiancaiamao @shenli @crazycs520 PTAL

**Comment by tiancaiamao** [comment]
> it will not wait for this down TiDB syncing the schema, then continue to change the TiDB schema to version 3

A schema change should meet either of the two conditions:
1. There is no network partition, all peers agree with a consistent cluster state
2. Wait a lease since last change, to guarantee that a schema won't change twice within a lease

If none of the condition is satisfied, how could it change to a newer version 3 ?

> So we need to stop the schema validator to prohibit the DML executing when we lost session of etcd.

We need not stop the schema validator, we should stop the version change in the owner instead. @winkyao

**Comment by shenli** [comment]
@winkyao Please add a test case.

**Comment by winkyao** [comment]
I'll add test cases.

**Comment by winkyao** [comment]
@tiancaiamao It is another issue, we can fix the issue you mentioned later, but not his PR.

File: ddl/fail_db_test.go [hunk 1]
```diff
@@ -197,3 +197,44 @@ func (s *testDBSuite) TestAddIndexFailed(c *C) {
...
+	}
+
+	c.Assert(s.dom.SchemaValidator.IsStarted(), IsFalse)
+	_, err := tk.Exec("insert into t values(1)")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[domain:3]Schema validator is stopped, it means TiDB is disconnected from the pd/etcd.")
+	s.dom.MockReloadFailed.SetValue(false)
+	// wait the schemaValidator is started.
+	for i := 0; i < 10; i++ {
+		time.Sleep(500 * time.Millisecond)
```
Comment (tiancaiamao) [inline ddl/fail_db_test.go]: Move this line after 234, and you can increase retry count and reduce sleep time.

File: domain/domain.go [hunk 2]
```diff
@@ -415,11 +415,26 @@ func (do *Domain) loadSchemaInLoop(lease time.Duration) {
...
+			// And version 1 is not consistent to version 3. So we need to stop the schema validator to prohibit the DML executing.
+			do.SchemaValidator.Stop()
 			err := do.mustRestartSyncer()
 			if err != nil {
 				log.Errorf("[ddl] reload schema in loop, schema syncer restart err %v", errors.ErrorStack(err))
 				break
 			}
+			// The schema maybe changed, must reload schema then the schema validator can restart.
+			err = do.mustReload()
+			if err != nil {
```
Comment (tiancaiamao) [inline domain/domain.go]: Is it always the case that the error is caused by "domain is closed" ?
If not, and `loadSchemaInLoop` exit, we'll in trouble.

File: domain/domain.go [hunk 3]
```diff
@@ -453,6 +468,28 @@ func (do *Domain) mustRestartSyncer() error {
...
+		}
+
+		// If the domain is closed, we returns immediately.
+		select {
+		case <-do.exit:
+			return errors.Trace(err)
+		default:
+		}
+
+		time.Sleep(time.Second)
```
Comment (tiancaiamao) [inline domain/domain.go]: So once we lost connection to etcd, the service unavailable time is at lease several seconds ?

File: domain/schema_validator.go [hunk 4]
```diff
@@ -166,7 +177,7 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
 	defer s.mux.RUnlock()
 	if !s.isStarted {
 		log.Infof("[domain-ddl] the schema validator stopped before checking")
-		return ResultFail
```
Comment (tiancaiamao) [inline domain/schema_validator.go]: When the returned result is `ResultFail`, the caller will sleep a while and retry.
After you change it to `ResultStopped`, what will happen? @winkyao

Comment (winkyao) [reply domain/schema_validator.go]: I know the caller will retry when the error is infoschemaChanged, but if the schemavalidator is stopped, shall we retry or return the error immediately?

File: domain/domain.go (hunk 3 above)
Comment (winkyao) [reply domain/domain.go]: Why do you think so? The service restore depends on when TiDB can reconnection to etcd.

File: domain/domain.go (hunk 2 above)
Comment (winkyao) [reply domain/domain.go]: return error always when "domain is closed". Please take a look in do.mustReload()

Comment (winkyao) [reply domain/domain.go]: Maybe I should change the return values to a boolean, indicate should we return the loop?

File: domain/schema_validator.go (hunk 4 above)
Comment (winkyao) [reply domain/schema_validator.go]: Done, we now will retry ErrSchemaValidatorStopped in session.isRetryableError. Please take a look. @tiancaiamao

Comment (winkyao) [reply domain/schema_validator.go]: Fine

File: domain/domain.go (hunk 3 above)
Comment (tiancaiamao) [reply domain/domain.go]: Could this retry duration be smaller?

Comment (winkyao) [reply domain/domain.go]: What is the proper retry duration do you think?

**Comment by tiancaiamao** [comment]
LGTM

**Comment by winkyao** [comment]
@crazycs520 @zimulala PTAL

**Pull request merged**